### PR TITLE
WT-8516 Patch CMake to support Evergreen testing

### PR DESCRIPTION
**Pull request by alisonfel**
This is a follow on ticket from WT-8418 with another set of small miscellaneous fixes and improvements, identified whilst integrating CMake with Evergreen (WT-8020). This PR should be the last major set of CMake patches to support WT-8020 (the next PR should be WT-8020 itself).
This PR includes the following patches:
- Extend CMake `define_build_mode` to accept libraries
- Unset `HAVE_BUILD_MODE_X` cache vars on error paths
- Fix ASan build mode flags for GCC builds
- Add a `Coverage` build mode
- Separate compiler versions between C & CXX for toolchain files
- Detect if `libclangfuzzer` is available before building the fuzz test.
- Only compile workgen if `ENABLE_PYTHON` is set

_To ease reviewing, each of the above changes are broken into individual feature commits._

### CODE REVIEW

File: cmake/configs/modes.cmake [hunk 1]
```diff
@@ -123,10 +123,15 @@ if(MSVC)
...
     set(asan_compiler_c_flag "/fsanitize=address")
     set(asan_compiler_cxx_flag "/fsanitize=address")
+    set(asan_lib_flags "")
 else()
-    set(asan_link_flags "-fsanitize=address")
     set(asan_compiler_c_flag "-fsanitize=address")
     set(asan_compiler_cxx_flag "-fsanitize=address")
+    set(asan_link_flags "-fsanitize=address")
+    if(GNU_C_COMPILER AND GNU_CXX_COMPILER)
+        set(asan_link_flags "${asan_link_flags}")
```
Comment (etienneptl) [inline cmake/configs/modes.cmake:132]: Why is this line necessary?

Comment (alisonfel) [reply cmake/configs/modes.cmake]: Oh whoops, good catch. That must've been left over from debugging. Removed in 544b10d.

**Review by etienneptl — APPROVED** [review]
LGTM, awesome work!

**Review by jiechenbo — APPROVED** [review]
LGTM!
